pennylane/gradients/general_shift_rules.py

```diff
@@ -378,6 +381,44 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
     return _combine_shift_rules(rules)
 
 
+def _copy_and_shift_params(tape, indices, shifts, multipliers, cast=False):
+    """Create a copy of a tape and of parameters, and set the new tape to the parameters
+    rescaled and shifted as indicated by ``indices``, ``multipliers`` and ``shifts``."""
+    all_ops = tape.circuit
+
+    for idx, shift, multiplier in zip(indices, shifts, multipliers):
+        _, op_idx, p_idx = tape.get_operation(idx)
+        op = (
+            all_ops[op_idx].obs
+            if isinstance(all_ops[op_idx], MeasurementProcess)
+            else all_ops[op_idx]
+        )
+
+        # Shift copied parameter
+        new_params = list(op.data)
+        new_params[p_idx] = new_params[p_idx] * qml.math.convert_like(multiplier, new_params[p_idx])
+        new_params[p_idx] = new_params[p_idx] + qml.math.convert_like(shift, new_params[p_idx])
+        if cast:
+            dtype = getattr(new_params[p_idx], "dtype", float)
+            new_params[p_idx] = qml.math.cast(new_params[p_idx], dtype)
+
+        # Create operator with shifted parameter and put into shifted tape
+        shifted_op = bind_new_parameters(op, new_params)

```

Comment:
Shouldn't `new_params = list(op.data)` and `shifted_op = bind_new_parameters(op, new_params)` happen before and after the `for` loop, respectively?

Reply:
No. The for loop iterates through all the parameter indices that that are going to be shifted, and then updates the operator corresponding to the trainable index by the shifted parameter using `bind_new_parameters`

Reply:
Oh yes, this makes sense. Sorry.